serialization/src/main/scala/sbt/serialization/SerializedValue.scala

```diff
@@ -22,116 +24,107 @@ import sbt.serialization.json.{
+      case lazyValue: LazyValue[_] =>
+        lazyValue.toJson.pickledValue.parsedValue
+    }
+
+  final override def equals(other: Any): Boolean =
+    other match {
+      case null => false
+      case sv: SerializedValue => toJValue.equals(sv.toJValue)
+      case _ => false
+    }
+
+  final override def hashCode(): Int =
+    toJValue.hashCode()
 }
 
 object SerializedValue {
   def apply[V](value: V)(implicit pickler: SPickler[V]): SerializedValue =
-    JsonValue[V](value)(pickler)
+    LazyValue[V](value, pickler)
 
-  /** Reconstitutes a SerialziedValue from a json string. */
+  /** Reconstitutes a SerializedValue from a json string. */
   def fromJsonString(value: String): SerializedValue =
-    JsonValue.fromJsonString(value)
-
-  // TODO - Expose fromBinaryBlob if/when we support binary.
-}
-
-// TODO - If this is meant to handle any kind of pickle format, it needs
-//        to encode the pickle format directly.
-//        i.e. it'd be nice if we had some kind of class that had a PickleFormat and Pickle
-//        if we ever wanted to support alternative SerializedValues.
-private[sbt] sealed trait SbtPrivateSerializedValue extends SerializedValue {
-  /**
-   * Returns this serialized value pickled into a json Tree.
-   *
-   * Note: This may be an expensive operation.
-   */
-  def toJson: JsonValue
-  /**
-   * Returns the serialzied value as a raw JSON string.
-   *
-   * Note: this may be an expensive operation.
-   */
-  def toJsonString: String
+    JsonValue(JSONPickle(value))
+
+  // TODO fromBinary if/when we add binary pickling
+
+  // this is in this file so it can use private JsonValue,
+  // but the public implicit version is in a trait elsewhere
+  private[sbt] object pickler extends SPickler[SerializedValue] with Unpickler[SerializedValue] {
+    val cheaterTag = implicitly[FastTypeTag[JValue]]
+    // TODO - This is super hacky mechanism to avoid issues w/ pinned types.
+    override val tag = cheaterTag.asInstanceOf[FastTypeTag[SerializedValue]]
+    def pickle(a: SerializedValue, builder: PBuilder): Unit = {
+      val json = a.toJValue
+      builder.hintTag(cheaterTag)
+      builder.hintStaticallyElidedType()
+      builder.beginEntry(json)
+      builder.endEntry()
+      //jsonPickler.pickle(spsv.toJson, builder)
+    }
+    def unpickle(tag: String, preader: PReader): Any = {
+      preader.hintTag(cheaterTag)
+      preader.hintStaticallyElidedType()
+      preader.beginEntry()
+      // TODO - Check beginEntry returns cheaterTag
+      val value = preader.readPrimitive().asInstanceOf[JValue]
+      preader.endEntry()
+      JsonValue(JSONPickle.fromJValue(value))
+    }
+  }
 }
-
-import sbt.serialization.json.JsonMethods._
 /** A value we have serialized as JSON */
-private[sbt] final case class JsonValue(pickledValue: JSONPickle) extends SbtPrivateSerializedValue {
+private final case class JsonValue(pickledValue: JSONPickle) extends SerializedValue {
+  require(pickledValue ne null)
```

Comment:
this isn't supposed to happen at all, it just gets you a more useful backtrace when JSONPickleFormat is messed up. I left it in just for future hackers on that. It's a bug if this triggers.